File: source/js/Auth/Clouded.Auth.Library/src/auth-management.client.ts
import { BaseManagementClient } from '@/base/base-management.client';
import { CloudedPromise } from '@/clouded-promise';

export class AuthManagementClient extends BaseManagementClient {
  private readonly organizationUrlSegment: string = '/v1/organizations';
  private readonly userUrlSegment: string = '/v1/users';
  private readonly roleUrlSegment: string = '/v1/roles';
  private readonly permissionUrlSegment: string = '/v1/permissions';
  private readonly machinesUrlSegment: string = '/v1/machines';

  constructor(apiUrl: string, apiKey: string, secretKey: string, cloudedKey?: string) {
    super(apiUrl, apiKey, secretKey, cloudedKey);
  }

  public machines(): CloudedPromise<any[]> {
    return this.getAll(this.machinesUrlSegment);
  }

  public machineById(id: number | string, withUsers = false): CloudedPromise<any> {
    return this.getById(this.machinesUrlSegment, id, { withUsers });
  }

  public machineCreate(input: any): CloudedPromise<any> {
    return this.create(this.machinesUrlSegment, input);
  }

  public machineUpdate(id: number | string, input: any): CloudedPromise<any> {
    return this.update(this.machinesUrlSegment, id, input);
  }

  public machineDelete(id: number | string): Promise<boolean> {
    return this.delete(this.machinesUrlSegment, id);
  }
  public organizations(): CloudedPromise<any[]> {
    return this.getAll(this.organizationUrlSegment);
  }

  public organizationById(id: number | string, withUsers = false): CloudedPromise<any> {
    return this.getById(this.organizationUrlSegment, id, { withUsers });
  }

  public organizationCreate(input: any): CloudedPromise<any> {
    return this.create(this.organizationUrlSegment, input);
  }

  public organizationUpdate(id: number | string, input: any): CloudedPromise<any> {
    return this.update(this.organizationUrlSegment, id, input);
  }

  public organizationDelete(id: number | string): Promise<boolean> {
    return this.delete(this.organizationUrlSegment, id);
  }

  public organizationUsersAdd(id: number | string, userIds: number[] | string[]): Promise<boolean> {
    return this.addOrRemove(this.organizationUrlSegment, id, 'users/add', userIds);
  }

  public organizationUsersRemove(
    id: number | string,
    userIds: number[] | string[],
  ): Promise<boolean> {
    return this.addOrRemove(this.organizationUrlSegment, id, 'users/remove', userIds);
  }

  public users(filter?: {
    search?: string;
    roleIds?: number[] | string[];
    permissionIds?: number[] | string[];
    organizationId?: number | string;
  }): CloudedPromise<any[]> {
    let queryParams = '';

    if (filter) {
      const { search, roleIds, permissionIds, organizationId } = filter;
      queryParams = [
        search && `search=${search}`,
        roleIds && roleIds.map((x) => `roleIds=${x}`).join('&'),
        permissionIds && permissionIds.map((x) => `permissionIds=${x}`).join('&'),
        organizationId && `organizationId=${organizationId}`,
      ]
        .filter((x) => x)
        .join('&');
    }

    return this.getAll(`${this.userUrlSegment}?${queryParams}`);
  }

  public usersFind(filter: {
    search?: string;
    roleIds?: number[] | string[];
    permissionIds?: number[] | string[];
    organizationId?: number | string;
  }): CloudedPromise<any[]> {
    const { search, roleIds, permissionIds, organizationId } = filter;
    const queryParams = [
      search && `search=${search}`,
      roleIds && roleIds.map((x) => `roleIds=${x}`).join('&'),
      permissionIds && permissionIds.map((x) => `permissionIds=${x}`).join('&'),
      organizationId && `organizationId=${organizationId}`,
    ]
      .filter((x) => x)
      .join('&');

    return this.httpClient
      .get(`${this.userUrlSegment}/find?${queryParams}`)
      .then((response) => response.data);
  }

  public userById(
    id: number | string,
    withRoles = false,
    withPermissions = false,
  ): CloudedPromise<any> {
    return this.getById(this.userUrlSegment, id, { withRoles, withPermissions });
  }

  public userByIdentity(
    identity: number | string,
    withRoles = false,
    withPermissions = false,
  ): CloudedPromise<any> {
    return this.httpClient
      .get(`${this.userUrlSegment}/by_identity/${identity}`, {
        params: { withRoles, withPermissions },
      })
      .then((response) => response.data);
  }

  public userCreate(input: any): CloudedPromise<any> {
    return this.create(this.userUrlSegment, input);
  }

  public userUpdate(id: number | string, input: any): CloudedPromise<any> {
    return this.update(this.userUrlSegment, id, input);
  }

  public userDelete(id: number | string): Promise<boolean> {
    return this.delete(this.userUrlSegment, id);
  }

  public userRolesAdd(
    id: number | string,
    roleIds: number[] | string[],
    organizationId?: number | string,
  ): Promise<boolean> {
    return this.addOrRemove(this.userUrlSegment, id, 'roles/add', {
      roleIds,
      organizationId,
    });
  }

  public userRolesRemove(
    id: number | string,
    roleIds: number[] | string[],
    organizationId?: number | string,
  ): Promise<boolean> {
    return this.addOrRemove(this.userUrlSegment, id, 'roles/remove', {
      roleIds,
      organizationId,
    });
  }

  public userPermissionsAdd(
    id: number | string,
    permissionIds: number[] | string[],
    organizationId?: number | string,
  ): Promise<boolean> {
    return this.addOrRemove(this.userUrlSegment, id, 'permissions/add', {
      permissionIds,
      organizationId,
    });
  }

  public userPermissionsRemove(
    id: number | string,
    permissionIds: number[] | string[],
    organizationId?: number | string,
  ): Promise<boolean> {
    return this.addOrRemove(this.userUrlSegment, id, 'permissions/remove', {
      permissionIds,
      organizationId,
    });
  }

  public roles(): CloudedPromise<any[]> {
    return this.getAll(this.roleUrlSegment);
  }

  public roleById(id: number | string): CloudedPromise<any> {
    return this.getById(this.roleUrlSegment, id);
  }

  public roleByIdentity(identity: number | string, withPermissions = false): CloudedPromise<any> {
    return this.httpClient
      .get(`${this.roleUrlSegment}/by_identity/${identity}`, {
        params: { withPermissions },
      })
      .then((response) => response.data);
  }

  public roleCreate(input: number | string): CloudedPromise<any> {
    return this.create(this.roleUrlSegment, input);
  }

  public roleUpdate(id: number | string, input: any): CloudedPromise<any> {
    return this.update(this.roleUrlSegment, id, input);
  }

  public roleDelete(id: number | string): Promise<boolean> {
    return this.delete(this.roleUrlSegment, id);
  }

  public rolePermissionsAdd(
    id: number | string,
    permissionIds: number | string[],
  ): Promise<boolean> {
    return this.addOrRemove(this.roleUrlSegment, id, 'permissions/add', permissionIds);
  }

  public rolePermissionsRemove(
    id: number | string,
    permissionIds: number | string[],
  ): Promise<boolean> {
    return this.addOrRemove(this.roleUrlSegment, id, 'permissions/remove', permissionIds);
  }

  public permissions(): CloudedPromise<any[]> {
    return this.getAll(this.permissionUrlSegment);
  }

  public permissionById(id: number | string): CloudedPromise<any> {
    return this.getById(this.permissionUrlSegment, id);
  }

  public permissionByIdentity(identity: number | string): CloudedPromise<any> {
    return this.httpClient
      .get(`${this.roleUrlSegment}/by_identity/${identity}`)
      .then((response) => response.data);
  }

  public permissionCreate(input: number | string): CloudedPromise<any> {
    return this.create(this.permissionUrlSegment, input);
  }

  public permissionUpdate(id: number | string, input: number | string): CloudedPromise<any> {
    return this.update(this.permissionUrlSegment, id, input);
  }

  public permissionDelete(id: number | string): Promise<boolean> {
    return this.delete(this.permissionUrlSegment, id);
  }
}

File: source/dotnet/Auth/Clouded.Auth.Client/AuthManagementClient.cs
using Clouded.Auth.Client.Base;
using Clouded.Auth.Shared;
using Clouded.Auth.Shared.Pagination;
using Clouded.Results;
using Flurl.Http;

namespace Clouded.Auth.Client;

public class AuthManagementClient(
    string apiUrl,
    string? apiKey = null,
    string? secretKey = null,
    string? cloudedKey = null
) : BaseManagementClient(apiUrl, apiKey, secretKey, cloudedKey)
{
    private const string OrganizationUrlSegment = $"v1/{RoutesConfig.OrganizationsRoutePrefix}";
    private const string UserUrlSegment = $"v1/{RoutesConfig.UsersRoutePrefix}";
    private const string RoleUrlSegment = $"v1/{RoutesConfig.RolesRoutePrefix}";
    private const string PermissionUrlSegment = $"v1/{RoutesConfig.PermissionsRoutePrefix}";
    private const string MachinesUrlSegment = $"v1/{RoutesConfig.MachinesRoutePrefix}";

    public Task<CloudedOkResult<Paginated<Dictionary<string, object?>>>> Organizations(
        int page = 0,
        int size = 50,
        CancellationToken cancellationToken = default
    ) => GetPaginatedAsync(OrganizationUrlSegment, page, size, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> Organization(
        object id,
        CancellationToken cancellationToken = default
    ) => GetByIdAsync(OrganizationUrlSegment, id, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> OrganizationCreate(
        Dictionary<string, object?> input,
        CancellationToken cancellationToken = default
    ) => CreateAsync(OrganizationUrlSegment, input, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> OrganizationUpdate(
        object id,
        Dictionary<string, object?> input,
        CancellationToken cancellationToken = default
    ) => UpdateAsync(OrganizationUrlSegment, id, input, cancellationToken);

    public Task<bool> OrganizationDelete(
        object id,
        CancellationToken cancellationToken = default
    ) => DeleteAsync(OrganizationUrlSegment, id, cancellationToken);

    public Task<bool> OrganizationUsersAdd(
        object id,
        object[] userIds,
        CancellationToken cancellationToken = default
    ) => AddOrRemoveAsync(OrganizationUrlSegment, id, "users/add", userIds, cancellationToken);

    public Task<bool> OrganizationUsersRemove(
        object id,
        object[] userIds,
        CancellationToken cancellationToken = default
    ) => AddOrRemoveAsync(OrganizationUrlSegment, id, "users/remove", userIds, cancellationToken);

    public Task<CloudedOkResult<Paginated<Dictionary<string, object?>>>> Users(
        int page = 0,
        int size = 50,
        CancellationToken cancellationToken = default
    ) => GetPaginatedAsync(UserUrlSegment, page, size, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> UserById(
        object id,
        CancellationToken cancellationToken = default
    ) => GetByIdAsync(UserUrlSegment, id, cancellationToken);

    public async Task<Dictionary<string, object>> UserByIdentity(
        object identity,
        CancellationToken cancellationToken = default
    )
    {
        var request = RestClient
            .Request()
            .AppendPathSegments(UserUrlSegment, "by_identity", identity);

        request = AddCloudedKeyHeader(request);
        request = AddMachineKeyHeader(request);

        return await request.GetJsonAsync<Dictionary<string, object>>(
            HttpCompletionOption.ResponseContentRead,
            cancellationToken
        );
    }

    public Task<CloudedOkResult<Dictionary<string, object?>>> UserCreate(
        Dictionary<string, object?> input,
        CancellationToken cancellationToken = default
    ) => CreateAsync(UserUrlSegment, input, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> UserUpdate(
        Dictionary<string, object?> input,
        CancellationToken cancellationToken = default
    ) => CreateAsync(UserUrlSegment, input, cancellationToken);

    public Task<bool> UserDelete(object id, CancellationToken cancellationToken = default) =>
        DeleteAsync(UserUrlSegment, id, cancellationToken);

    public Task<bool> UserRolesAdd(
        object id,
        object[] roleIds,
        CancellationToken cancellationToken = default
    ) => AddOrRemoveAsync(UserUrlSegment, id, "roles/add", roleIds, cancellationToken);

    public Task<bool> UserRolesRemove(
        object id,
        object[] roleIds,
        CancellationToken cancellationToken = default
    ) => AddOrRemoveAsync(UserUrlSegment, id, "roles/remove", roleIds, cancellationToken);

    public Task<bool> UserPermissionsAdd(
        object id,
        object[] permissionIds,
        CancellationToken cancellationToken = default
    ) => AddOrRemoveAsync(UserUrlSegment, id, "permissions/add", permissionIds, cancellationToken);

    public Task<bool> UserPermissionsRemove(
        object id,
        object[] permissionIds,
        CancellationToken cancellationToken = default
    ) =>
        AddOrRemoveAsync(
            UserUrlSegment,
            id,
            "permissions/remove",
            permissionIds,
            cancellationToken
        );

    public Task<CloudedOkResult<Paginated<Dictionary<string, object?>>>> Roles(
        int page = 0,
        int size = 50,
        CancellationToken cancellationToken = default
    ) => GetPaginatedAsync(RoleUrlSegment, page, size, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> Role(
        object id,
        CancellationToken cancellationToken = default
    ) => GetByIdAsync(RoleUrlSegment, id, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> RoleCreate(
        Dictionary<string, object?> input,
        CancellationToken cancellationToken = default
    ) => CreateAsync(RoleUrlSegment, input, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> RoleUpdate(
        Dictionary<string, object?> input,
        CancellationToken cancellationToken = default
    ) => CreateAsync(RoleUrlSegment, input, cancellationToken);

    public Task<bool> RoleDelete(object id, CancellationToken cancellationToken = default) =>
        DeleteAsync(RoleUrlSegment, id, cancellationToken);

    public Task<bool> RolePermissionsAdd(
        object id,
        object[] permissionIds,
        CancellationToken cancellationToken = default
    ) => AddOrRemoveAsync(RoleUrlSegment, id, "permissions/add", permissionIds, cancellationToken);

    public Task<bool> RolePermissionsRemove(
        object id,
        object[] permissionIds,
        CancellationToken cancellationToken = default
    ) =>
        AddOrRemoveAsync(
            RoleUrlSegment,
            id,
            "permissions/remove",
            permissionIds,
            cancellationToken
        );

    public Task<CloudedOkResult<Paginated<Dictionary<string, object?>>>> Permissions(
        int page = 0,
        int size = 50,
        CancellationToken cancellationToken = default
    ) => GetPaginatedAsync(PermissionUrlSegment, page, size, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> Permission(
        object id,
        CancellationToken cancellationToken = default
    ) => GetByIdAsync(PermissionUrlSegment, id, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> PermissionCreate(
        Dictionary<string, object?> input,
        CancellationToken cancellationToken = default
    ) => CreateAsync(PermissionUrlSegment, input, cancellationToken);

    public Task<CloudedOkResult<Dictionary<string, object?>>> PermissionUpdate(
        Dictionary<string, object?> input,
        CancellationToken cancellationToken = default
    ) => CreateAsync(PermissionUrlSegment, input, cancellationToken);

    public Task<bool> PermissionDelete(object id, CancellationToken cancellationToken = default) =>
        DeleteAsync(PermissionUrlSegment, id, cancellationToken);
}
